File: leetcode_problems/python/lc704_binary_search.py
class Solution:
    def is_good(self, val: int, target: int) -> bool:
        return val <= target

    def search(self, nums: list[int], target: int) -> int:
        low, high = 0, len(nums)
        while high - low > 1:
            mid = low + (high - low) // 2
            if self.is_good(nums[mid], target):
                low = mid
            else:
                high = mid

        return low if nums[low] == target else -1


if __name__ == "__main__":
    nums = [-1, 0, 3, 5, 9, 12]
    target = 3

    print(Solution().search(nums, target))

File: leetcode_problems/cpp/lc704_binary_search.cpp
#include <vector>
#include <iostream>
using std::vector;

class Solution
{
public:
    bool is_good(int value, int target)
    {
        return value <= target;
    }

    int search(vector<int> &nums, int target)
    {
        int low = 0;
        int high = nums.size();
        while (high - low > 1)
        {
            int mid = low + (high - low) / 2;
            if (is_good(nums[mid], target))
            {
                low = mid;
            }
            else
            {
                high = mid;
            }
        }
        return nums[low] == target ? low : -1;
    }
};

int main()
{
    vector<int> nums = {-1, 0, 3, 5, 9, 12};
    int target = 9;
    Solution sol;
    int res = sol.search(nums, target);
    std::cout << res << std::endl;
    return 0;
}